### DISCUSSION TITLE
Is there a way not to use auto-imports at all?

### DISCUSSION THREAD

**Discussion by antoinerey** [nuxt/nuxt#21956] [Questions]
**We're building a medium/large scale application with Nuxt 3, and we would like to drop the auto-imports feature completely, is it possible?**

Before jumping to code examples, let me clarify that we are aware of disabling implicit auto-imports, and explicitly importing from #imports and #app instead. This is not what we are looking for. We would like to both disable implicit auto imports, and not explicitly import anything from # imports (#imports, #app, #build, #vue-router, etc.).

---

We are building several modules providing composables and helpers, and we currently rely on the `addImports` Nuxt function to make those composables accessible from consumer applications.

```ts
// auth-module/composables.ts
import { useState } from '#app'

export function useAuth() {
  // ...
}

// auth-module/module.ts
export default defineNuxtModule({
  setup() {
    addImports({ name: 'useAuth', from: resolve(runtimeDir, './composables') })
  }
})
```

On the consumer side, we then register this module, and retrieve the `useAuth` composable from `#imports`.

```ts
// consumer application
import { useAuth } from '#imports'
```

So, we tried to change the import not to use # imports anymore and ended up with the following.

```ts
// auth-module/composables.ts
import { useState } from 'nuxt/app'           // #app -> nuxt/app

export function useAuth() {
  // ...
}

// consumer application
import { useAuth } from '@acme/auth-module'   // #imports -> @acme/auth-module
```

---

However, we eventually hit some issues with Nuxt internal code. It seems that some code from `node_modules/nuxt/dist/app` also relies on # imports internally. Below are a couple of examples, but we found about 60 occurrences, in 25 files.

```ts
// node_modules/nuxt/dist/app/composables/preload.js
import { renderJsonPayloads } from "#build/nuxt.config.mjs";

// node_modules/nuxt/dist/app/composables/chunk.js
import { useNuxtApp } from "#app/nuxt";
```

**Given that Nuxt internal code also relies on # imports, is it safe or even possible not to rely on # imports inside our own code, and instead explicitly import stuff from package names as usual?**

**Comment by Hebilicious** [comment]
If I'm not mistaken, you can achieve something like this with custom aliases. I've always prefixed them with `#`, but this should work too ? 

```ts
    nuxt.hook("nitro:config", (nitroConfig) => {
      nitroConfig.alias = nitroConfig.alias || {}
      nitroConfig.alias[`@acme/auth-module`] = resolve("./runtime/server")
    })
    nuxt.options.alias[`@acme/auth-module`] = resolve("./runtime/client")
```
https://github.com/Hebilicious/nuxt-module-template/blob/main/packages/my-module/src/module.ts

**Reply by antoinerey** [reply]
That would allow me to import stuff as if it was an npm dependency indeed. 👍  

However, it does not fully answer to all my needs. Code located in `./runtime/server` and `./runtime/client` would still import Nuxt internal helpers (`useAsyncData`, `useRuntimeConfig`, etc) from `#imports`, which in turn causes troubles with tooling (component unit testing, etc).

What I would like to do is _never_ import anything from virtual imports (`#imports`, `#app`, etc).

**Reply by Hebilicious** [reply]
May I ask what problems do you encounter with `#imports` exactly ?

**Reply by antoinerey** [reply]
Sure! 🤗 

The main issues I have are all about surrounding tooling, mainly test runners and IDEs.

- Importing from virtual imports in tests, or testing code that relies on virtual imports is quite hacky to say the least. Vitest does not understand them, and so we have to provide their members using a different way. Right now, we rely mostly on mocking the `#imports` module thanks to Vitest mocking capabilities. It works but it also makes our test suites quite fragile. Every time we use a new virtual import member, we have to make sure a fake version is provided into the tests. I tend to see that as implementation details leaking to the tests.

- On the IDEs side, we often end up reloading VSCode (or other) because it can't properly find declaration files for virtual imports. If we do not touch modules boundaries (aka. do not regenerate anything), it stays stable. However, as soon as we start generating new `.nuxt/*` files, there are high chances they do not get properly loaded by the IDEs and that we have to reload it.

All in all, I'm not sure what problems virtual imports solve that standard ESM imports would not. I tend to see them as unnecessary complexity, and thus question their existence in the first place. Until this practice becomes way more common in the industry, I'm afraid that configuring tooling around them will be tricky at least, and will require more time and energy than anticipated.

Hoping that this answer clarifies my position a little bit. 🤗

**Reply by Hebilicious** [reply]
Virtual imports are aliases to real directories, you can configure your tsconfig like this https://github.com/Hebilicious/nuxt-module-template/blob/main/tsconfig.json, which is something necessary for module development. 
Configuring vitest properly with Nuxt should work better than trying to get rid of virtual imports. Look at this [vitest.config](https://github.com/nuxt/nuxt/blob/main/vitest.config.ts), you can specify alias there. 
I think we need more documentation around these parts though.

For the IDE loading, there might be a bug, if we can reproduce it we should be able to fix it.

**Reply by antoinerey** [reply]
> Virtual imports are aliases to real directories, you can configure your tsconfig like this [Hebilicious/nuxt-module-template@main/tsconfig.json](https://github.com/Hebilicious/nuxt-module-template/blob/main/tsconfig.json), which is something necessary for module development.

Ah! I might be missing something then.

I can't find any proper file that would resolve from `#imports`. Unlike in your template where you map `#app` to `resolve('./packages/nuxt/dist/app/index')`, I don't have such single file to map to `#imports`.

What is eventually resolved from `#imports` comes from all over the place. A bit from Vue, a bit from Nuxt, a bit from our custom code, etc. To the best of my knowledge, there is no single file on my filesystem that gather all of that. Am I wrong?

**Reply by TheAlexLichter** [reply]
You can check `.nuxt/imports.d.ts` to see what can be imported from there.

**Reply by antoinerey** [reply]
Mmmh. I've tried configuring Vitest aliases so that `#imports` resolves to the `.nuxt/imports.d.ts` from my module playground, but that does not work.

```ts
// vitest.config.ts

export default {
  resolve: {
    alias: {
      '#import': path.resolve(__dirname, './playground/.nuxt/imports.d.ts'),
    }
  }
}
```

After setting this, I got two errors. 

```
Error: Failed to load url vue-demi (resolved id: vue-demi) in /omitted/playground/.nuxt/imports.d.ts. Does the file exist?

Error: Failed to load url #app (resolved id: #app) in /omitted/playground/.nuxt/imports.d.ts. Does the file exist?
```

So, I've added both as aliases as well.

```ts
// vitest.config.ts

export default {
  resolve: {
    alias: {
      '#import': path.resolve(__dirname, './playground/.nuxt/imports.d.ts'),

      // I've added these ones.
	  'vue-demi': resolve(__dirname, '../../node_modules/nuxt/dist/app/compat/vue-demi.d.ts'),
	  '#app': resolve(__dirname, '../../node_modules/nuxt/app.d.ts'),
    }
  }
}
```

Which solved the issues above, but raised a new one.

```
TypeError: Package import specifier "#build/app.config.mjs" is not defined in package /omitted/node_modules/nuxt/package.json imported from /omitted/node_modules/nuxt/dist/app/config.js
```

It seems that this one can't be solved by simply throwing an alias at it. I've added one, but I still have the same issue.

```ts
// vitest.config.ts

export default {
  resolve: {
    alias: {
	  'vue-demi': resolve(__dirname, '../../node_modules/nuxt/dist/app/compat/vue-demi.d.ts'),
	  '#app': resolve(__dirname, '../../node_modules/nuxt/app.d.ts'),
      '#import': path.resolve(__dirname, './playground/.nuxt/imports.d.ts'),

      // I've added this one.
	  '#build/app.config.mjs': resolve(__dirname, './playground/.nuxt/app.config.mjs'),
    }
  }
}
```

On a related note. I've tried to replicate the `unimport` Nuxt configuration in Vitest, or to use `vite-tsconfig-paths` already in the past. Every time, I ended up on this `#build/app.config.mjs` issue, which is exactly why I've started this discussion in the first place.

**Reply by Hebilicious** [reply]
Have you tried to specify `#build` instead of `#build/app.config.mjs`?

This is the config that has been working for me :

```
  "#imports": ["./packages/my-module/node_modules/nuxt/app.d.ts"],
  "#app": ["./packages/my-module/node_modules/nuxt/dist/app"]
```
I think we should create an issue from this because the documentation is lacking when it comes to your use case.

**Reply by antoinerey** [reply]
> Have you tried to specify #build instead of #build/app.config.mjs?

I double-checked after your message. And the results are identical unfortunately. I still get the following error message.

```
TypeError: Package import specifier "#build/app.config.mjs" is not defined in package /omitted/node_modules/nuxt/package.json imported from /omitted/node_modules/nuxt/dist/app/config.js
```
